[DeviceSanitizer] Support out-of-bounds on private memory

AllanZyne opened the pull request:
- Add runtime support for private memory, and use "detect_privates" option to control it
- When the shadow memory for private or local memory is allocated failed, the checking will be disabled automatically
- Fix WA for local work group size since "urKernelGetSuggestedLocalWorkSize" is added
- Set the default level of sanitizer logger to "warning"

LLVM: https://github.com/intel/llvm/pull/13935

AllanZyne commented:
Hi @oneapi-src/unified-runtime-maintain, @zhaomaosu, @yingcong-wu and @wenju-he, can you please review this PR?
Thank you very much!

zhaomaosu reviewed (approved):
lgtm

wenju-he commented on source/loader/layers/sanitizer/asan_libdevice.hpp:
```diff
@@ -82,6 +82,10 @@ struct LaunchInfo {
 
+// Based on the observation, only the last 24 bits of the address of the private
+// variable have changed, we use 31 bits(2G) to be safe.
+constexpr std::size_t ASAN_PRIVATE_SIZE = 0x7fffffffULL + 1;
```
2G is probably too big, how about just use 24 bits?

AllanZyne replied:
We should set ASAN_PRIVATE_SIZE as large as possible, if not, it will produce unpredictable wrong error report.

wenju-he replied:
is there a reproducer that produces a wrong report if we use 24 bits?

AllanZyne replied:
I tried a create a kernel that has large private memory but failed at "PI_OUT_OF_RESOURCES" when the private memory size is more than 8MB without asan.
Besides, if asan is enabled, the maximum size of private memory on user side should be small.

Therefore, I think it's fine to use 24 bits now.

wenju-he replied:
Thank you for the test. Using 24bit lgtm. Though it will still cause allocating a huge memory, which may result in OOM, when number of workgroups is large. I guess we probably need to reduce ASAN_PRIVATE_SIZE further in the future.

AllanZyne replied:
I think we can also prevent from OOM by limiting the detecting range of workgroup

pbalcer commented on source/loader/layers/sanitizer/asan_interceptor.cpp:
```diff
@@ -202,6 +202,12 @@ SanitizerInterceptor::SanitizerInterceptor() {
+    KV = Options->find("detect_privates");
+    if (KV != Options->end()) {
+        auto Value = KV->second.front();
+        cl_DetectPrivates = Value == "1" || Value == "true" ? true : false;
```
nit: you have now at least two ways to check for true. Maybe create a helper?

yingcong-wu commented on source/loader/layers/sanitizer/asan_interceptor.cpp:
```diff
@@ -760,12 +766,8 @@ ur_result_t SanitizerInterceptor::prepareLaunch(
         if (LaunchInfo.LocalWorkSize.empty()) {
-            LaunchInfo.LocalWorkSize.reserve(3);
+            LaunchInfo.LocalWorkSize.resize(3);
             // FIXME: This is W/A until urKernelSuggestGroupSize is added
```
urKernelSuggestGroupSize has been added now, do you plan to address this FIXME?

AllanZyne replied:
Nice catch! Added.

yingcong-wu reviewed (approved):
lgtm

wenju-he reviewed (approved):
LGTM

wenju-he reviewed (approved):
LGTM

AllanZyne commented:
Hi @oneapi-src/unified-runtime-maintain, @zhaomaosu, @yingcong-wu and @wenju-he, can you please review this PR again?
I added some detail changes since last review (also added description in PR).
Thank you very much!

yingcong-wu reviewed (approved):
lgtm

zhaomaosu reviewed (approved):
lgtm

callumfare commented:
@AllanZyne This has conflicts now that #1677 has merged. If you could fix them then I'll make sure this gets merged next. Thanks!

AllanZyne commented:
> @AllanZyne This has conflicts now that #1677 has merged. If you could fix them then I'll make sure this gets merged next. Thanks!

Fixed. Thanks very much!
